keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1349 @@
+
+Syntax has always been a magnet for arguments in our industry.
+We expect this to be no exception.
+Reasonable people will disagree with our choices,
+which is fine, but we should hold a strong opinion on this.
+
+YAML is a very large and complex specification.
+We expect that we will find places where KYAML may be less readable than conventional YAML.
+We should revisit this specification when that happens.
+
+### Risks and Mitigations
+
+Risk: Our encoder produces some output which is not valid YAML.
+Mitigation: That’s a bug. It’s hard to imagine a case that can’t be fixed.
+
+Risk: The result could be net worse than plain YAML.
+Mitigation: We will produce many examples and include them in the review process. We will solicit feedback from various groups on the result.
+
+Risk: This gives a false sense of safety, but doesn’t SOLVE  the problems with YAML, just avoids them.
+Mitigation: Docs?
+
+Risk: Like all APIs, people will come to depend on the details, which may prevent us from changing them.
+Mitigation: Be as precise as possible where we are confident, but leave room where we are not.
+
+## Design Details
+
+Creating an encoder for this format is not very technically challenging,
+but there are many details to consider.
+
+### Packaging
+
+<<[UNRESOLVED where it lives]>>
+
+The simplest thing to do is to write a new implementation of
+k8s.io/cli-runtime/pkg/printers.ResourcePrinter (in
+staging/src/k8s.io/cli-runtime/pkg/printers). However, this isn’t the most
+obviously reusable place for it.
+
+A more ecosystem-friendly solution is to write a new function or package under
+sigs.k8s.io/yaml](sigs.k8s.io/yaml) which can be used by other tools.
+
+<<[/UNRESOLVED]>>
+
+### Tooling
+
+k/k has a CI verifier (hack/verify-yamlfmt.sh) which looks for YAML files and
+ensures they are formatted “correctly”. The only meaning of “correct” is “block
+style”, and it corrupts KYAML input today (aside: it also does not handle
+multi-documents).  If we intend to convert examples, we need to fix this.
+
+We propose to add flags to it which specify to use “conventional” YAML or KYAML
+output, and set the default to KYAML (or to change the parent script to always
+specify KYAML).
+
+This tooling will preserve comments.
+
+#### How?
+
+Today we have forked go-yaml.v2 AND go-yaml.v3 into sigs.k8s.io/yaml.  While
+KYAML is an output format, our tooling goals require that we get access to a
+YAML AST, which may require some dirty tricks.  If we ever change YAML
+libraries (e.g.
+[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml)), we may
+need different dirty tricks.
+
+Both sigs.k8s.io/yaml/go-yaml.v2 and v3 offer a way to decode YAML into an AST,
+which we can render as KYAML.  Neither, however, offers a way to encode an
+arbitrary Go object as the equivalent YAML AST.  To do that we may have to
+encode the object to plain YAML, then decode it to AST, then render it as
```

Comment:
I would expect us to encode to json using the standard json encoder, then decode, then transform / render as kyaml. That gets us all the json encoding behaviors for ~free.

Reply:
It also loses comments.  If you run `yamlfmt --kyaml` (or whatever we end up with) on a file, it should reflow it to KYAML but retain the comments.

Reply:
I think there's two entry points to the pipeline:
1. In-memory object (the PrintObject entry point).
2. Yaml bytes (which includes json bytes) we want to transform to kyaml (from files or arbitrary inputs)

If we handle entry point 1 by serializing to JSON, we'll perfectly match all JSON semantics and behavior (and will preserve struct order for typed inputs and sort keys for unstructured / map inputs, just like -o json), and can then route that output to entry point 2, reflowing it into kyaml format.

Reply:
Maybe I am missing something.

Case 1: From a Go object.
- Call `json.Marshal()` on it (preserves field order).
- Call `json.Unmarshal()` to decode to a map (loses field order).
- Call `kyaml.Marshal()` 

?  Am I misunderstanding?  Or maybe you are assuming that we can convert JSON to an AST?  I guess I could try through `json.Decoder.Token()` and re-encode that -- is that what you meant?

Case 2a: From JSON bytes.
- Use Decoder as suggested above

Case 2b: From YAML bytes.
- Decode to `yaml.Node` (retains comments)
- Render to kyaml.

That leaves me with 2 paths still?

Reply:
I was envisioning:

Case 1: From a Go object.

Call json.Marshal() on it (preserves field order). The output should be passable to case 2, preserving order.

Case 2: From JSON or YAML bytes.

Decode to yaml.Node (retains comments)
Render to kyaml.

Reply:
Oh, or maybe you meant:

Case 1: From a Go object.
- Call `json.Marshal()` on it (preserves field order).
- Call `yaml.Unmarshal(*yaml.Node)` on the JSON bytes
- Render Node -> KYAML

Case 2: From JSON bytes.
- Call `yaml.Unmarshal(*yaml.Node)` on the JSON bytes
- Render Node -> KYAML

Case 3: From YAML bytes.
- Call `yaml.Unmarshal(*yaml.Node)` on the YAML bytes (preserve comments)
- Render Node -> KYAML

That could work.

Possible problem: JSON "streams" are not supported by YAML as multi-documents.  That's PROBABLY fine?

Reply:
Mid-air collision.  That may be possible.  I had "forgotten" that i can pass JSON bytes to yaml library (unless it is a multi-document stream)

Reply:
Another advantage of first invoking the JSON encoder: it protects against types and data structures which cannot be encoded (keys that are not supported, cycles). Otherwise that has to be re-implemented by the KYAML encoder.

Reply:
Detailed in the next KEP push